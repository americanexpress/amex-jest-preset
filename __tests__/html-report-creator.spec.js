/*
 * Copyright (c) 2017 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

const merge = require('lodash/merge');
const mkdirp = require('mkdirp');
const fs = require('fs');

function buildMockResult(mockResult = {}) {
  const defaultMockResult = {
    numFailedTestSuites: 0,
    numFailedTests: 0,
    numPassedTestSuites: 1,
    numPassedTests: 1,
    numPendingTestSuites: 0,
    numPendingTests: 0,
    numRuntimeErrorTestSuites: 0,
    numTotalTestSuites: 1,
    numTotalTests: 1,
    startTime: 1497400767268,
    success: true,
    testResults: [
      { console: [],
        perfStats: {
          end: 1497401364334,
          start: 1497401364037,
        },
        testFilePath: '/path/to/__tests__/test.spec.js',
        testResults: [
          {
            failureMessages: [],
            duration: 4,
            fullName: 'test should work',
            status: 'passed',
          },
        ],
      },
    ],
  };

  return merge({}, defaultMockResult, mockResult);
}

function setupTest() {
  jest.doMock('mkdirp', () => ({ sync: jest.fn() }));
  fs.writeFileSync = jest.fn();
  process.cwd = jest.fn(() => '/path/to/something');

  // eslint-disable-next-line global-require
  return require('../html-report-creator');
}

describe('html-report-creator', () => {
  const originalFsWriteFileSync = fs.writeFileSync;
  const originalCwd = process.cwd();
  const originalMkdirpSync = mkdirp.sync;
  const originalJestTestReportPath = process.env.JEST_TEST_REPORT_PATH;

  beforeEach(() => {
    jest.resetModules();
    jest.resetAllMocks();
    process.env.JEST_TEST_REPORT_PATH = 'path/to/something/test-report.html';
  });

  it('should render correctly if there are test execution errors in a suite', () => {
    const createHtmlReport = setupTest();
    const mockResult = buildMockResult(
      {
        testResults: [
          { testExecError: {
            message: 'error!',
            stack: 'stacktrace for error',
          },
          },
        ],
      }
    );

    createHtmlReport(mockResult);
    expect(
      fs.writeFileSync.mock.calls[0][1]
    ).toMatchSnapshot();
  });

  it('should not display pending tests', () => {
    const createHtmlReport = setupTest();
    const mockResult = buildMockResult(
      {
        numPassedTestSuites: 0,
        numPendingTestSuites: 1,
        numPassedTests: 0,
        numPendingTests: 1,
        testResults: [
          {
            testResults: [{ status: 'pending' }],
          },
        ],
      }
    );

    createHtmlReport(mockResult);
    expect(
      fs.writeFileSync.mock.calls[0][1]
    ).toMatchSnapshot();
  });

  it('should correctly render passing tests', () => {
    const createHtmlReport = setupTest();
    const mockResult = buildMockResult();

    createHtmlReport(mockResult);
    expect(
      fs.writeFileSync.mock.calls[0][1]
    ).toMatchSnapshot();
  });

  it('should correctly render failing tests', () => {
    const createHtmlReport = setupTest();
    const mockResult = buildMockResult(
      {
        numPassedTestSuites: 0,
        numFailedTestSuites: 1,
        numPassedTests: 0,
        numFailedTests: 1,
        testResults: [
          {
            testResults: [{ status: 'failed', failureMessages: ['This test failed yo!'] }],
          },
        ],
      }
    );

    createHtmlReport(mockResult);
    expect(
      fs.writeFileSync.mock.calls[0][1]
    ).toMatchSnapshot();
  });

  it('should ignore null plainFailureMessages', () => {
    const createHtmlReport = setupTest();
    const mockResult = buildMockResult(
      {
        numPassedTestSuites: 0,
        numFailedTestSuites: 1,
        numPassedTests: 0,
        numFailedTests: 1,
        testResults: [
          {
            testResults: [{ status: 'failed', failureMessages: [null] }],
          },
        ],
      }
    );

    createHtmlReport(mockResult);
    expect(
      fs.writeFileSync.mock.calls[0][1]
    ).toMatchSnapshot();
  });

  it('should render failure message as a link if test is for an image snapshot', () => {
    const createHtmlReport = setupTest();
    const mockResult = buildMockResult(
      {
        numPassedTestSuites: 0,
        numFailedTestSuites: 1,
        numPassedTests: 0,
        numFailedTests: 1,
        testResults: [
          {
            testResults: [{ status: 'failed', failureMessages: ['Image snapshot failed.\nSee diff for details: /path/to/__image_snapshots__/__diff_output__/test-diff.png'] }],
          },
        ],
      }
    );

    createHtmlReport(mockResult);
    expect(
      fs.writeFileSync.mock.calls[0][1]
    ).toMatchSnapshot();
  });

  it('should write output to default path if JEST_TEST_REPORT_PATH does not exist', () => {
    delete process.env.JEST_TEST_REPORT_PATH;
    const createHtmlReport = setupTest();
    const mockResult = buildMockResult();

    createHtmlReport(mockResult);
    expect(fs.writeFileSync.mock.calls[0][0])
      // eslint-disable-next-line no-useless-escape
      .toMatch(/[\\\/]path[\\\/]to[\\\/]something[\\\/]test-results[\\\/]test-report.html/);
  });

  it('should write output to JEST_TEST_REPORT_PATH if variable exists', () => {
    const customTestReportPath = 'custom/path/to/test-report.html';
    process.env.JEST_TEST_REPORT_PATH = customTestReportPath;
    const createHtmlReport = setupTest();
    const mockResult = buildMockResult();

    createHtmlReport(mockResult);
    expect(fs.writeFileSync.mock.calls[0][0]).toBe(customTestReportPath);
  });

  afterAll(() => {
    fs.writeFileSync = originalFsWriteFileSync;
    process.cwd = originalCwd;
    mkdirp.sync = originalMkdirpSync;
    if (originalJestTestReportPath) {
      process.env.JEST_TEST_REPORT_PATH = originalJestTestReportPath;
    } else {
      delete process.env.JEST_TEST_REPORT_PATH;
    }
  });
});
