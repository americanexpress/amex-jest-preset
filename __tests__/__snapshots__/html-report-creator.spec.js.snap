// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html-report-creator should correctly render failing tests 1`] = `
"<?xml version="1.0"?>
<html>
  <head>
    <meta charset="utf-8"/>
    <title>Jest test results</title>
    <style type="text/css">
  html, body {
    font-family: sans-serif;
    font-size: 1rem;
  }
  body {
    padding: 1rem 1rem;
  }
  .summary {
    color: #999;
    margin-bottom: 1em;
  }
  .suite-info {
    padding-bottom: 1em;
    font-weight: bold;
    font-size: 0.9rem;
    color: black;
  }
  .suite-table {
    width: 100%;
    margin-bottom: 1em;
  }
  .suite-table tr.passed {
    background-color: mintcream;
    color: green;
  }
  .suite-table tr.failed {
    background-color: LavenderBlush;
    color: red;
  }
  .suite-table td {
    font-size: 0.85rem;
    border-bottom: 1px solid #aaa;
    vertical-align: top;
    padding: 0.5rem;
  }
  .suite-table span.result {
    float: right;
  }
</style>
  </head>
  <div class="summary">
    <div>
    Test Suites: 0 passed,
    1 failed,
    0 skipped,
    1 total
  </div>
    <div>
    Tests: 0 passed,
    1 failed,
    0 skipped,
    1 total
  </div>
  </div>
  <div class="suite-info">
      /path/to/__tests__/test.spec.js
      (0.297s)
    </div>
  <table class="suite-table">
    <tr class="failed">
      <td>
        ✘ test should work
        <span class="result">failed in 0.004s</span>
        <div>
          <pre>This test failed yo!</pre>
        </div>
      </td>
    </tr>
  </table>
</html>"
`;

exports[`html-report-creator should correctly render passing tests 1`] = `
"<?xml version="1.0"?>
<html>
  <head>
    <meta charset="utf-8"/>
    <title>Jest test results</title>
    <style type="text/css">
  html, body {
    font-family: sans-serif;
    font-size: 1rem;
  }
  body {
    padding: 1rem 1rem;
  }
  .summary {
    color: #999;
    margin-bottom: 1em;
  }
  .suite-info {
    padding-bottom: 1em;
    font-weight: bold;
    font-size: 0.9rem;
    color: black;
  }
  .suite-table {
    width: 100%;
    margin-bottom: 1em;
  }
  .suite-table tr.passed {
    background-color: mintcream;
    color: green;
  }
  .suite-table tr.failed {
    background-color: LavenderBlush;
    color: red;
  }
  .suite-table td {
    font-size: 0.85rem;
    border-bottom: 1px solid #aaa;
    vertical-align: top;
    padding: 0.5rem;
  }
  .suite-table span.result {
    float: right;
  }
</style>
  </head>
  <div class="summary">
    <div>
    Test Suites: 1 passed,
    0 failed,
    0 skipped,
    1 total
  </div>
    <div>
    Tests: 1 passed,
    0 failed,
    0 skipped,
    1 total
  </div>
  </div>
  <div class="suite-info">
      /path/to/__tests__/test.spec.js
      (0.297s)
    </div>
  <table class="suite-table">
    <tr class="passed">
      <td>
        ✔ test should work
        <span class="result">passed in 0.004s</span>
      </td>
    </tr>
  </table>
</html>"
`;

exports[`html-report-creator should ignore null plainFailureMessages 1`] = `
"<?xml version="1.0"?>
<html>
  <head>
    <meta charset="utf-8"/>
    <title>Jest test results</title>
    <style type="text/css">
  html, body {
    font-family: sans-serif;
    font-size: 1rem;
  }
  body {
    padding: 1rem 1rem;
  }
  .summary {
    color: #999;
    margin-bottom: 1em;
  }
  .suite-info {
    padding-bottom: 1em;
    font-weight: bold;
    font-size: 0.9rem;
    color: black;
  }
  .suite-table {
    width: 100%;
    margin-bottom: 1em;
  }
  .suite-table tr.passed {
    background-color: mintcream;
    color: green;
  }
  .suite-table tr.failed {
    background-color: LavenderBlush;
    color: red;
  }
  .suite-table td {
    font-size: 0.85rem;
    border-bottom: 1px solid #aaa;
    vertical-align: top;
    padding: 0.5rem;
  }
  .suite-table span.result {
    float: right;
  }
</style>
  </head>
  <div class="summary">
    <div>
    Test Suites: 0 passed,
    1 failed,
    0 skipped,
    1 total
  </div>
    <div>
    Tests: 0 passed,
    1 failed,
    0 skipped,
    1 total
  </div>
  </div>
  <div class="suite-info">
      /path/to/__tests__/test.spec.js
      (0.297s)
    </div>
  <table class="suite-table">
    <tr class="failed">
      <td>
        ✘ test should work
        <span class="result">failed in 0.004s</span>
        <div/>
      </td>
    </tr>
  </table>
</html>"
`;

exports[`html-report-creator should not display pending tests 1`] = `
"<?xml version="1.0"?>
<html>
  <head>
    <meta charset="utf-8"/>
    <title>Jest test results</title>
    <style type="text/css">
  html, body {
    font-family: sans-serif;
    font-size: 1rem;
  }
  body {
    padding: 1rem 1rem;
  }
  .summary {
    color: #999;
    margin-bottom: 1em;
  }
  .suite-info {
    padding-bottom: 1em;
    font-weight: bold;
    font-size: 0.9rem;
    color: black;
  }
  .suite-table {
    width: 100%;
    margin-bottom: 1em;
  }
  .suite-table tr.passed {
    background-color: mintcream;
    color: green;
  }
  .suite-table tr.failed {
    background-color: LavenderBlush;
    color: red;
  }
  .suite-table td {
    font-size: 0.85rem;
    border-bottom: 1px solid #aaa;
    vertical-align: top;
    padding: 0.5rem;
  }
  .suite-table span.result {
    float: right;
  }
</style>
  </head>
  <div class="summary">
    <div>
    Test Suites: 0 passed,
    0 failed,
    1 skipped,
    1 total
  </div>
    <div>
    Tests: 0 passed,
    0 failed,
    1 skipped,
    1 total
  </div>
  </div>
  <div class="suite-info">
      /path/to/__tests__/test.spec.js
      (0.297s)
    </div>
  <table class="suite-table"/>
</html>"
`;

exports[`html-report-creator should render correctly if there are test execution errors in a suite 1`] = `
"<?xml version="1.0"?>
<html>
  <head>
    <meta charset="utf-8"/>
    <title>Jest test results</title>
    <style type="text/css">
  html, body {
    font-family: sans-serif;
    font-size: 1rem;
  }
  body {
    padding: 1rem 1rem;
  }
  .summary {
    color: #999;
    margin-bottom: 1em;
  }
  .suite-info {
    padding-bottom: 1em;
    font-weight: bold;
    font-size: 0.9rem;
    color: black;
  }
  .suite-table {
    width: 100%;
    margin-bottom: 1em;
  }
  .suite-table tr.passed {
    background-color: mintcream;
    color: green;
  }
  .suite-table tr.failed {
    background-color: LavenderBlush;
    color: red;
  }
  .suite-table td {
    font-size: 0.85rem;
    border-bottom: 1px solid #aaa;
    vertical-align: top;
    padding: 0.5rem;
  }
  .suite-table span.result {
    float: right;
  }
</style>
  </head>
  <div class="summary">
    <div>
    Test Suites: 1 passed,
    0 failed,
    0 skipped,
    1 total
  </div>
    <div>
    Tests: 1 passed,
    0 failed,
    0 skipped,
    1 total
  </div>
  </div>
  <div class="suite-info">
      /path/to/__tests__/test.spec.js
      (0.297s)
    </div>
  <table class="suite-table">
    <tr class="failed">
      <td>
        <div>Test suite failed to run:</div>
        <pre>error!
stacktrace for error</pre>
      </td>
    </tr>
  </table>
</html>"
`;

exports[`html-report-creator should render failure message as a link if test is for an image snapshot 1`] = `
"<?xml version="1.0"?>
<html>
  <head>
    <meta charset="utf-8"/>
    <title>Jest test results</title>
    <style type="text/css">
  html, body {
    font-family: sans-serif;
    font-size: 1rem;
  }
  body {
    padding: 1rem 1rem;
  }
  .summary {
    color: #999;
    margin-bottom: 1em;
  }
  .suite-info {
    padding-bottom: 1em;
    font-weight: bold;
    font-size: 0.9rem;
    color: black;
  }
  .suite-table {
    width: 100%;
    margin-bottom: 1em;
  }
  .suite-table tr.passed {
    background-color: mintcream;
    color: green;
  }
  .suite-table tr.failed {
    background-color: LavenderBlush;
    color: red;
  }
  .suite-table td {
    font-size: 0.85rem;
    border-bottom: 1px solid #aaa;
    vertical-align: top;
    padding: 0.5rem;
  }
  .suite-table span.result {
    float: right;
  }
</style>
  </head>
  <div class="summary">
    <div>
    Test Suites: 0 passed,
    1 failed,
    0 skipped,
    1 total
  </div>
    <div>
    Tests: 0 passed,
    1 failed,
    0 skipped,
    1 total
  </div>
  </div>
  <div class="suite-info">
      /path/to/__tests__/test.spec.js
      (0.297s)
    </div>
  <table class="suite-table">
    <tr class="failed">
      <td>
        ✘ test should work
        <span class="result">failed in 0.004s</span>
        <div>
          <a href=" /path/to/__image_snapshots__/__diff_output__/test-diff.png"/>
          <pre>Image snapshot failed.
See diff for details: /path/to/__image_snapshots__/__diff_output__/test-diff.png</pre>
        </div>
      </td>
    </tr>
  </table>
</html>"
`;
